File: src/python/k_variance_plot.py
#!/usr/local/bin/python

"""
Stores the data for plotting the evolution of the objective function
value over increasing value of k over the time for different settings
"""
from MarkovChain import *
import pandas as pd
import os
from plot_hubway_stations import *

DATA_DIR = "/Users/harshal/Projects/markov_traffic/Plots_data/"

dataframe_rows = []

def get_objective_evolution(mc, problem, object_distribution,
                            method, k, method_name):
    rows = mc.get_evolution(method, k)
    for row in rows:
        row['problem'] = problem
        row['object_distribution'] = object_distribution
        row['time'] = mc.current_time
        row['method'] = method_name

    global dataframe_rows
    dataframe_rows = dataframe_rows + rows

if __name__ == "__main__":
    num_objects, transition_matrix, G, object_distribution = get_mc_attributes(duration=600)
    mc1 = MCNodeObjectives(len(G.nodes()), num_objects, 0, transition_matrix, object_distribution, G)
    max_k = 50
    # Node objective evolution

    # 1. Uniform object distribution
    get_objective_evolution(mc1, 'node', 'actual', mc1.smart_greedy, max_k, 'greedy')

    get_objective_evolution(mc1, 'node', 'actual', mc1.random_nodes, max_k, 'random')

    get_objective_evolution(mc1, 'node', 'actual',
                mc1.highest_in_probability_nodes, max_k, 'highest_in_probability')

    get_objective_evolution(mc1, 'node', 'actual',
                mc1.highest_in_degree_centrality_nodes, max_k, 'highest_in_degree_centrality')

    get_objective_evolution(mc1, 'node', 'actual',
                mc1.highest_closeness_centrality_nodes, max_k, 'highest_closeness_centrality')

    get_objective_evolution(mc1, 'node', 'actual',
                mc1.highest_betweenness_centrality_nodes, max_k, 'highest_betweenness_centrality')

    get_objective_evolution(mc1, 'node', 'actual',
                mc1.highest_closeness_centrality_nodes, max_k, 'highest_item_nodes')

    # Edge objective evolution
    mc1 = MCEdgeObjectives(len(G.nodes()), num_objects, 0, transition_matrix, object_distribution, G)

    # 1. Uniform object distribution
    get_objective_evolution(mc1, 'edge', 'actual', mc1.smart_greedy, max_k, 'greedy')

    get_objective_evolution(mc1, 'edge', 'actual', mc1.random_edges, max_k, 'random')

    get_objective_evolution(mc1, 'edge', 'actual', mc1.highest_probability_edges, max_k, 'high_probability')

    get_objective_evolution(mc1, 'edge', 'actual', mc1.highest_betweenness_centrality_edges, max_k, 'highest_betweenness_centrality')

    get_objective_evolution(mc1, 'edge', 'actual', mc1.dp_algorithm, max_k, 'dynamic_program')

    get_objective_evolution(mc1, 'edge', 'actual', mc1.highest_probability_edges, max_k, 'highest_item_edges')

    # Create a dataframe
    df = pd.DataFrame(dataframe_rows)

    # Export the data to a csv in DATA_DIR
    df.to_csv(DATA_DIR + "k_objective_evolution_hubway.csv.gz", sep=",", header=True, index=False, compression='gzip')

    # Call the R script to make the plots
    # os.system("~/Projects/markov_traffic/src/R/k_variance_plot.r")

File: src/R/k_variance_plot.r
#! /usr/local/bin/RScript

library(data.table)
library(ggplot2)
library(ggthemes)

# Set working directories
setwd("~/Projects/markov_traffic/")
figures.dir = "~/Projects/markov_traffic/Plots/"

# Read data
dd = data.table(read.delim("./Plots_data/k_objective_evolution_hubway.csv.gz",
                           sep=","))
dd[, object_distribution := as.character(object_distribution)]
dd[, method := as.character(method)]
dd[, problem := as.character(problem)]


# Node k evolution plot for time=0
node.dd = dd[problem == "node"]
node.dd = node.dd[time == 0]
max_k = max(node.dd$k)
p1 = ggplot(data=node.dd, aes(x=k, y=objective_value))
p1 = p1 + geom_line(aes(color=method, group=method))
# p1 = p1 + geom_point()
p1 = p1 + scale_x_continuous(name="k", breaks=seq(0,max_k,10))
p1 = p1 + scale_y_continuous(name="Node Monitoring Objective")
p1 = p1 + facet_wrap(~object_distribution, nrow=1)
p1 = p1 + theme_bw()
p1 = p1 + theme(strip.background = element_blank())
p1
ggsave(paste0(figures.dir, "hubway_nodes_k_evolution.pdf"), w=12, h=4)

# Edge k evolution plot for time=0
edge.dd = dd[problem == "edge"]
edge.dd = edge.dd[time == 0]
max_k = max(edge.dd$k)
p2 = ggplot(data=edge.dd, aes(x=k, y=objective_value))
p2 = p2 + geom_line(aes(color=method, group=method))
# p2 = p2 + geom_point()
p2 = p2 + scale_x_continuous(name="k", breaks=seq(0,max_k,10))
p2 = p2 + scale_y_continuous(name="Edge Monitoring Objective")
p2 = p2 + facet_wrap(~object_distribution, nrow=1)
p2 = p2 + theme_bw()
p2 = p2 + theme(strip.background = element_blank())
p2
ggsave(paste0(figures.dir, "hubway_edges_k_evolution.pdf"), w=12, h=4)
